java/arcs/core/storage/handle/CollectionImpl.kt

```diff
@@ -26,9 +27,10 @@ typealias SetBase<T> = Handle<CrdtSet.Data<T>, CrdtSet.IOperation<T>, Set<T>>
  * It provides methods that can generate the appropriate operations to send to a
  * backing storage proxy.
  */
-class CollectionImpl<T : Referencable>(
+class CollectionImpl<T : Referencable, Q>(
     name: String,
-    storageProxy: SetProxy<T>
+    storageProxy: SetProxy<T>,
+    private val refinement: Refinement<T, Q>?
```

Comment:
How does this look when the collection does not have an associated refinement? It looks like you will still need to parameterise the collection by the type Q, even though it doesn't affect anything. Is there any way around that?

Reply:
I don't think there's a way to have optional type parameters in Kotlin.

We could have a `QueryableCollectionImpl` as the underlying implementation of `CollectionImpl`, where `CollectionImpl` is an alias for `QueryableCollectionImpl<T, Unit>`?

Reply:
Would it be possible for QueryableCollectionImpl to subclass CollectionImpl?

Reply:
I think this would be painful, as all the types that refer to CollectionImpl would need to be duplicated for QueryableCollectionImpl. As CollectionImpl isn't intended to be visible to the user I'd much prefer the complexity to reside here.

I think this would also set us up to need to add other <feature>ableCollectionImpl classes and there's no obvious way (to me) to compose these. Is there a way to support composition in the Kotlin type system: e.g. `CollectionImpl<Queryable<Q>, Deletable, ...>`?

Reply:
Not that I know of.

Reply:
Have updated, hiding the types needed for the query. Shane pointed out that we should be checking the types with the code generated from the refinements itself anyway (the handle shouldn't care).